cou/steps/plan.py

```diff
@@ -416,58 +429,105 @@ def _generate_control_plane_plan(
     :rtype: list[UpgradePlan]
     """
     principal_upgrade_plan = create_upgrade_group(
-        apps=apps,
+        apps=[app for app in apps if app.is_subordinate is False],
         description="Control Plane principal(s) upgrade plan",
         target=target,
         force=force,
-        filter_function=lambda app: not isinstance(app, SubordinateBase),
     )
 
     subordinate_upgrade_plan = create_upgrade_group(
-        apps=apps,
+        apps=[app for app in apps if app.is_subordinate],
```

Comment:
It's not an real issue for us, but like this, we are iterating over apps multiple times.

Reply:
No need to change it in this PR, but we eventually should change it and remove / improve `create_upgrade_group`, which really unintuitive.

Reply:
> It's not an real issue for us, but like this, we are iterating over apps multiple times.

Is it different from what `filter` in the end of the day does?

Reply:
Yes, if you pass a generator here and use that generator in function, it would be fine. But like this you are iterating over all apps and then again iterating only on filtered apps.